engine/src/core/error/expected.h

```diff
@@ -0,0 +1,108 @@
+//
+// Created by grzegorz on 18.09.19.
+//
+
+#pragma once
+
+#include <utility>
+#include "unexpected.h"
+#include "bald_assert.h"
+
+
+namespace Bald {
+    template<typename E, typename U>
+    class expected {
+        union {
+            E valid;
+            U invalid;
+        };
+        bool _isOk = true;
+    public:
+        constexpr expected() { new(&valid) E(); }
+
+        constexpr expected(const E& rhs) { new(&valid) E(rhs); }
+
+        constexpr expected(const unexpected<U>& rhs) :
+                _isOk(false) { new(&invalid) U(rhs.value); }
+
+        constexpr expected(E&& rhs) { new(&valid) E(std::forward<E>(rhs)); }
+
+        constexpr expected(unexpected<U>&& rhs) :
+                _isOk(false) { new(&invalid) U(std::forward<U>(rhs.value)); }
+
+        constexpr expected(const expected<E, U>& rhs) :
+                _isOk(rhs._isOk) {
+            if (rhs._isOk) new(&valid) E(rhs.valid);
+            else new(&invalid) U(rhs.invalid);
+        }
+
+        constexpr expected(expected<E, U>&& rhs) noexcept :
+                _isOk(rhs._isOk) {
+            if (rhs._isOk) new(&valid) E(std::move(rhs.valid));
+            else new(&invalid) U(std::move(rhs.invalid));
+        }
+
+        constexpr expected<E, U>& operator=(const expected<E, U>& other) {
```

Comment:
Be consistent. If you use "rhs" to refer to rand hand side of the assignment, keep it consistent across the implementation. Same applies to methods defined below.

Reply:
You provide an implementation for assignment operator with "expected" as an argument for both L-val and R-val reference, which is good since you also provide appropriate constructors. However, there is no assignment operator provided for L-val/R-val for the "unexpected", but the constructors for these types are provided. Please elaborate on this.